### PR TITLE
5836compare blocks

### PR DESCRIPTION
**Pull request by jdevcs**
## Description

https://github.com/web3/web3.js/issues/5836

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist for 1.x:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` with success.
- [ ] I have tested the built `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

## Checklist for 4.x:

- [ ] I have selected the correct 4.x base branch.
- [ ] Within the description, the feature or issue is discussed in detail or an issue is linked.   
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added any required tests for the changes I made 
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `yarn` successfully
- [ ] I ran `yarn lint` successfully
- [ ] I ran `yarn build:web` successfully
- [ ] I ran `yarn test:unit` successfully
- [ ] I ran `yarn test:integration` successfully
- [ ] I ran `compile:contracts` successfully
- [ ] I have tested my code.
- [ ] I have updated the corresponding `CHANGELOG.md` file in the packages I have edited.

### CODE REVIEW

**Review by avkos — APPROVED** [review]

File: packages/web3-utils/src/validation.ts [hunk 1]
```diff
@@ -109,56 +109,34 @@ export const isTopicInBloom = isTopicInBloomValidator;
...
-		// b !== a, thus a < b
-		return -1;
-	}
-	if (blockB === 'earliest' || blockB === 0) {
-		// b !== a, thus a > b
-		return 1;
-	}
-	if (blockA === 'latest' || blockA === 'safe') {
-		if (blockB === 'pending' || blockB === 'latest') {
+
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/src/validation.ts:127]: I think we can still compare `earliest` and  `0` with any tag or number. Because, `earliest` is a special tag that is always equal to `0`. And the following code is good enough for that:
```suggestion

	if (blockA === 'earliest' || blockA === 0) {
		// b !== a, thus a < b
		return -1;
	}
	if (blockB === 'earliest' || blockB === 0) {
		// b !== a, thus a > b
		return 1;
	}
```

However, some comments and tests would need to be updated accordingly.

File: docs/docs/guides/web3_migration_guide/web3_utils_migration_guide.md [hunk 2]
```diff
@@ -62,3 +62,15 @@ isHex('-0x'); // in 1.x used to return `true`. But changed in 4.x to return `fal
 isHexStrict('-0x'); // in 1.x used to return `true`. But changed in 4.x to return `false`
 // `false`
 ```
+
+## Other functions
+
+`compareBlockNumbers` now accepts either both block tags or both block numbers for comparison as parameters. The only exception is comparison of block tag `earliest` with 0.
```
Comment (Muhammad-Altabba) [inline docs/docs/guides/web3_migration_guide/web3_utils_migration_guide.md]:
```suggestion
`compareBlockNumbers` now accepts either both block tags or both block numbers for comparison as parameters. The only exception is the comparison of block tag `earliest` or 0 with any other.
```

File: packages/web3-utils/test/fixtures/validation.ts [hunk 3]
```diff
@@ -28,41 +28,43 @@ export const compareBlockNumbersValidData: [[Numbers, Numbers], number][] = [
...
-	[['', 'pending'], new InvalidBlockError('')],
-	[[22, 'finalized'], new InvalidBlockError('Cannot compare finalized tag with 22')],
-	[['finalized', 22], new InvalidBlockError('Cannot compare finalized tag with 22')],
-	[['genesis', 'finalized'], new InvalidBlockError('Genesis tag not supported')],
+	[['pending', 'unknown'], errorObj],
+	[['', 'pending'], errorObj],
+	[[22, 'finalized'], errorObj],
+	[['finalized', 22], errorObj],
+	[['earliest', 2], errorObj],
+	[[2, 'earliest'], errorObj],
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/test/fixtures/validation.ts]:
```suggestion
```

File: packages/web3-utils/test/fixtures/validation.ts [hunk 4]
```diff
@@ -28,41 +28,43 @@ export const compareBlockNumbersValidData: [[Numbers, Numbers], number][] = [
...
+	[['latest', 110], errorObj],
+	[[222, 'latest'], errorObj],
+	[['pending', 230], errorObj],
+	[[10000, 'pending'], errorObj],
+	[['latest', BigInt(1)], errorObj],
+	[['pending', BigInt(1)], errorObj],
+	[['safe', 0], errorObj],
+	[[0, 'safe'], errorObj],
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/test/fixtures/validation.ts:67]:
```suggestion
```

File: packages/web3-utils/test/fixtures/validation.ts [hunk 5]
```diff
@@ -28,41 +28,43 @@ export const compareBlockNumbersValidData: [[Numbers, Numbers], number][] = [
 	[[1, BigInt(1)], 0],
 	[[1, BigInt(2)], -1],
 	[[2, BigInt(1)], 1],
-
 	[['earliest', 0], 0],
+	[[0, 'earliest'], 0],
+	[['earliest', 'earliest'], 0],
 	[['pending', 'pending'], 0],
 	[['latest', 'latest'], 0],
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/test/fixtures/validation.ts:35]: Bring back some deleted tests as they will be valid again if the suggestion to code was applied.
```suggestion
	[['latest', 'latest'], 0],
	[['earliest', 2], -1],
	[[2, 'earliest'], 1],
```

File: packages/web3-utils/test/fixtures/validation.ts [hunk 6]
```diff
@@ -28,41 +28,43 @@ export const compareBlockNumbersValidData: [[Numbers, Numbers], number][] = [
...
 	[['earliest', 'pending'], -1],
 	[[BigInt('9007199254740992'), BigInt('9007199254740991')], 1],
 	[[13532346, 13532300], 1],
 	[['pending', 'latest'], 1],
-	[['latest', 0], 1],
-	[['latest', BigInt(1)], 1],
-	[['pending', 0], 1],
-	[['pending', BigInt(1)], 1],
 	[['safe', 'safe'], 0],
 	[['earliest', 'safe'], -1],
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/test/fixtures/validation.ts:41]: Bring back some deleted tests as they will be valid again if the suggestion to code was applied.
```suggestion
	[['earliest', 'safe'], -1],
	[['safe', 0], 1],
	[[0, 'safe'], -1],
```

File: packages/web3-utils/test/fixtures/validation.ts [hunk 7]
```diff
@@ -28,41 +28,43 @@ export const compareBlockNumbersValidData: [[Numbers, Numbers], number][] = [
...
-	[['safe', 0], 1],
-	[[0, 'safe'], -1],
 	[['safe', 'pending'], -1],
 	[['pending', 'safe'], 1],
 	[['finalized', 'finalized'], 0],
 	[['earliest', 'finalized'], -1],
-	[['finalized', 0], 1],
 	[['finalized', 'pending'], -1],
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/test/fixtures/validation.ts:46]: Bring back some deleted tests as they will be valid again if the suggestion to code was applied.
```suggestion
    [['earliest', 'finalized'], -1],
	[['finalized', 0], 1],
	[['finalized', 'pending'], -1],
	[[0, 'finalized'], -1],
```

**Review by Muhammad-Altabba — COMMENTED** [review]
LGTM.
However, I have a suggestion to still be able to compare 0 and `earliest` to any other tag or block number. As `earliest` is always `0`. I added suggestions for this.

File: packages/web3-utils/src/validation.ts (hunk 1 above)
Comment (jdevcs) [reply packages/web3-utils/src/validation.ts:121]: > I think we can still compare earliest and 0 with any tag or number. 
for earliest tag vs num:
what if user passes Number.MIN_VALUE , or run it on some chain with 0 blocks mined and latest/safe/finalized is also 0 (slim chances but still edge case that cannot be managed until we have block number for latest, safe, finalized,..etc).

Comment (jdevcs) [reply packages/web3-utils/src/validation.ts:127]: > we can still compare earliest and 0 with any tag or number.

1. we can compare `earliest` tag vs numbers `but with sanity check in place`: updated

2. 0 with `any tag` : I think not as in some edge cases as: if some one running local node for testing with block mining not set to time interval but set to some # of txs ( like geth insta mode ). In this case safe, latest etc block tags can be at 0 . so tags vs num case tag can be high or same and cant be compared until tag num is fetched first.

File: packages/web3-utils/test/fixtures/validation.ts (hunk 4 above)
Comment (jdevcs) [reply packages/web3-utils/test/fixtures/validation.ts:67]: shared reason above, in some slim edge case safe can be 0.

File: packages/web3-utils/test/fixtures/validation.ts (hunk 5 above)
Comment (jdevcs) [reply packages/web3-utils/test/fixtures/validation.ts:35]: added case for earliest.

File: docs/docs/guides/web3_migration_guide/web3_utils_migration_guide.md (hunk 2 above)
Comment (jdevcs) [reply docs/docs/guides/web3_migration_guide/web3_utils_migration_guide.md]: updated

File: packages/web3-utils/test/fixtures/validation.ts (hunk 3 above)
Comment (jdevcs) [reply packages/web3-utils/test/fixtures/validation.ts]: updated

File: packages/web3-utils/test/fixtures/validation.ts (hunk 6 above)
Comment (jdevcs) [reply packages/web3-utils/test/fixtures/validation.ts:41]: shared reason above, in some slim edge case safe can be 0.

File: packages/web3-utils/test/fixtures/validation.ts (hunk 7 above)
Comment (jdevcs) [reply packages/web3-utils/test/fixtures/validation.ts:46]: shared reason above, in some slim edge case finalized can be 0.

File: packages/web3-utils/src/validation.ts (hunk 1 above)
Comment (Muhammad-Altabba) [reply packages/web3-utils/src/validation.ts:127]: Good catch!

File: packages/web3-utils/CHANGELOG.md [hunk 8]
```diff
@@ -83,3 +83,6 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 -   Added support of `safe` and `finalized` block tags (#5823)
 
+### Changed
+
+-   `compareBlockNumbers` function now only supports comparison of both blocktags params ( except `earliest` vs 0) or both block number params
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/CHANGELOG.md]: I think it is now "(except `earliest` vs numbers)"

```suggestion
-   `compareBlockNumbers` function now only supports the comparison of either if both are blocktags (except `earliest` vs numbers) or if both are block number
```

**Review by Muhammad-Altabba — APPROVED** [review]
Great!
I have just a tiny suggestion.
And if you could please run `yarn changelog sync` to update the main CHANGELOG.md file.
